scrapy/core/http2/protocol.py

```diff
@@ -0,0 +1,183 @@
+import logging
+
+from h2.config import H2Configuration
+from h2.connection import H2Connection
+from h2.events import (
+    ConnectionTerminated, DataReceived, ResponseReceived,
+    StreamEnded, StreamReset, WindowUpdated
+)
+from twisted.internet.protocol import connectionDone, Protocol
+
+from scrapy.core.http2.stream import Stream
+from scrapy.http import Request
+
+LOGGER = logging.getLogger(__name__)
+
+
+class H2ClientProtocol(Protocol):
+    # TODO:
+    #  1. Check for user-agent while testing
+    #  2. Add support for cookies
+    #  3. Handle priority updates (Not required)
+    #  4. Handle case when received events have StreamID = 0 (applied to H2Connection)
+    #  1 & 2:
+    #   - Automatically handled by the Request middleware
+    #   - request.headers will have 'Set-Cookie' value
+
+    def __init__(self):
+        config = H2Configuration(client_side=True, header_encoding='utf-8')
+        self.conn = H2Connection(config=config)
+
+        # Address of the server we are connected to
+        # these are updated when connection is successfully made
+        self.destination = None
+
+        # ID of the next request stream
+        # Following the convention made by hyper-h2 each client ID
+        # will be odd.
+        self.next_stream_id = 1
+
+        # Streams are stored in a dictionary keyed off their stream IDs
+        self.streams = {}
+
+        # Boolean to keep track the connection is made
+        # If requests are received before connection is made
+        # we keep all requests in a pool and send them as the connection
+        # is made
+        self.is_connection_made = False
+        self._pending_request_stream_pool = []
+
+    def _stream_close_cb(self, stream_id: int):
+        """Called when stream is closed completely
+        """
+        self.streams.pop(stream_id, None)
+
+    def _new_stream(self, request: Request):
+        """Instantiates a new Stream object
+        """
+        stream_id = self.next_stream_id
+        self.next_stream_id += 2
+
+        stream = Stream(
+            stream_id=stream_id,
+            request=request,
+            connection=self.conn,
+            write_to_transport=self._write_to_transport,
```

Comment:
Any reason to use this callback instead of giving the stream direct access to `self.transport`? It looks like `self.conn` and `self.transport` is all the stream needs from the protocol to write into the transport, and you are already passing `self.conn`.

Reply:
Yeah so the only of `self.transport` in `Stream` instance would be to write data into transport. So, instead of creating the same function I thought maybe passing the the `HTTP2ClientProtocol` would be better. Also, in case there are some edge cases we can directly change `HTTP2ClientProtocol._write_to_transport` to reflect changes everywhere